odiglet/pkg/ebpf/manager.go

```diff
@@ -0,0 +1,371 @@
+	defer close(m.done)
+
+	runLoopCtx, stop := context.WithCancel(ctx)
+	m.stop = stop
+
+	// main event loop for handling instrumentations
+	for {
+		select {
+		case <-runLoopCtx.Done():
+			m.Logger.Info("stopping Odiglet instrumentation manager")
+			for pid, details := range m.DetailsByPid {
+				err := details.Inst.Close(ctx)
+				if err != nil {
+					m.Logger.Error(err, "failed to close instrumentation", "pid", pid)
+				}
+				delete(m.DetailsByPid, pid)
+				// probably shouldn't remove instrumentation instance here
+				// as this flow is happening when Odiglet is shutting down
+			}
+			return
+		case e := <-m.ProcEvents:
+			switch e.EventType {
+			case detector.ProcessExecEvent:
+				m.Logger.Info("detected new process", "pid", e.PID, "cmd", e.ExecDetails.CmdLine)
+				err := m.handleProcessExecEvent(runLoopCtx, e)
+				// ignore the error if no instrumentation factory is found,
+				// as this is expected for some language and sdk combinations
+				if err != nil && !errors.Is(err, ErrNoInstrumentationFactory) {
+					m.Logger.Error(err, "failed to handle process exec event")
+				}
+			case detector.ProcessExitEvent:
+				m.cleanInstrumentation(runLoopCtx, e.PID)
+			}
+		case configUpdate := <-m.configUpdates:
+			if configUpdate.config == nil {
+				m.Logger.Info("received nil config update, skipping")
+			}
+			for _, sdkConfig := range configUpdate.config.Spec.SdkConfigs {
+				err := m.applyInstrumentationConfigurationForSDK(runLoopCtx, configUpdate.workloadKey, &sdkConfig)
+				if err != nil {
+					m.Logger.Error(err, "failed to apply instrumentation configuration")
+				}
+			}
+		}
+	}
+}
+
+func (m *Manager) Stop() {
+	m.stop()
+	<-m.done
+}
+
+func (m *Manager) cleanInstrumentation(ctx context.Context, pid int) {
+	details, found := m.DetailsByPid[pid]
+	if !found {
+		m.Logger.V(3).Info("no instrumentation found for exiting pid", "pid", pid)
+		return
+	}
+
+	m.Logger.Info("cleaning instrumentation resources", "pid", pid)
+
+	err := details.Inst.Close(ctx)
+	if err != nil {
+		m.Logger.Error(err, "failed to close instrumentation")
+	}
+
+	// remove instrumentation instance
+	if err = m.Client.Delete(ctx, &odigosv1.InstrumentationInstance{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      instance.InstrumentationInstanceName(details.Pod.Name, pid),
+			Namespace: details.Pod.Namespace,
+		},
+	}); err != nil && !apierrors.IsNotFound(err) {
+		m.Logger.Error(err, "error deleting instrumentation instance", "pod", details.Pod.Name, "pid", pid)
+	}
+
+	delete(m.DetailsByPid, pid)
+}
+
+func (m *Manager) handleProcessExecEvent(ctx context.Context, e detector.ProcessEvent) error {
+	if _, found := m.DetailsByPid[e.PID]; found {
+		// this can happen if we have multiple exec events for the same pid (chain loading)
+		// TODO: better handle this?
+		// this can be done by first closing the existing instrumentation,
+		// and then creating a new one
+		m.Logger.Info("instrumentation already exists for pid", "pid", e.PID)
```

Comment:
Don't we need to reload instrumentation if we get another exec event on a process? If I am not mistaken, exec basically replaces the entire process memory but keep the same PID

Reply:
@blumamir  I improved the log message as you suggested.
@edeNFed  I guess the most common scenario is fork and exec - which will result in a new PID from the fork. An exec can be used by itself - a bash script calling `exec` is an example where that will happen.
Since the runtime-detector is currently configured to only pass exec events for processes that are alive longer than a second, I think we will filter a lot of theses cases - that is the reason I didn't include the full handling of this case here (reloading the instrumentation). Do you think we should add handling for it?